### PR TITLE
Add manual_soc_max: a discharge-direction ceiling sibling to manual_soc

### LINKED ISSUE #1578: FR - GivEnergy battery maintenance mode to better assist with SOC recalculations
**Is your feature request related to a problem? Please describe.**
GivEnergy batteries are known for the occasional SOC issue if batteries do not frequently get below 50% and stay idle for 30 minutes or longer under than threshold, I'd like to have an option toggle on/off that will force Predbat to on a weekly basis dump the entire battery just before a charge slot.

**Describe the solution you'd like**
 I'd like to have an option toggle on/off that will force Predbat to on a weekly basis dump the entire battery just before a charge slot.

**Describe alternatives you've considered**

Manually execute my above idea which is someitmes hard to remember with a growing family :).

**Additional context**
None

### PR DESCRIPTION
**Pull request by chalfontchubby**
## Summary

`select.predbat_manual_soc` already lets an automation set a **minimum** SoC floor for a specific time - Predbat plans charging to reach it by the deadline. There was no equivalent for the opposite direction: forcing the battery **down** to a ceiling by a deadline, which is what #1578 actually needs - a periodic calibration discharge finishing right before a known cheap import slot (e.g. an Octopus Intelligent Go midnight slot), so the BMS can re-anchor its SoC estimate before recharging cheaply.

Design discussion (including why the naive `select.predbat_manual_soc` "just set a lower target" approach doesn't work - it's documented as a minimum, not an exact/maximum target) happened on the issue and in-session; this PR is the resulting implementation.

**Why this should give better results than the community's existing workaround** (a fixed-duration forced-export automation, see #1578's thread): because the ceiling is a real cost signal integrated into the normal optimiser search rather than a blunt forced block, Predbat is free to use the discharged energy against load or export depending on whichever is actually cheapest at each moment, and stays coordinated with everything else already in the plan (car charging, existing charge/export windows). A fixed force-export block can't do either of those.

## Design

`manual_soc_max` / `manual_soc_max_value` mirror `manual_soc` / `manual_soc_value` exactly, just inverted:

- `fetch.py`: merges into a new per-minute `all_active_keep_max` dict (mirrors `all_active_keep`), using `min()` across sources instead of `max()` since it's a ceiling.
- `prediction.py`: the optimiser treats being **above** the ceiling at a "kept" minute as a cost penalty priced at the **export rate** (mirrors the floor's penalty, which prices being below it at the **import rate** - see the "pretend the excess should have been exported" comment).
- **Conflict handling**: a ceiling below the floor at the same minute is a contradiction (most likely a stale override left over from a previous automation run). The floor wins as the safety-relevant constraint, and the conflicting ceiling is dropped with a `Warn:` log line rather than handing the optimiser two penalties pulling opposite ways.

### The C++ kernel

`prediction.py`'s hot loop has a C++ mirror (`prediction_kernel.cpp`) for performance, with its own documented parity rule: any behavioural change must be mirrored there and both `PK_PARITY_REVISION`/`KERNEL_PARITY_REVISION` bumped, so a stale prebuilt binary is rejected at load time (falls back to the Python engine) rather than silently disagreeing with it. Done here - both bumped from 10 to 11.

Verified locally:
- Rebuilt the kernel (`build_kernel.sh`) - compiles clean with `-Wall -Werror`
- Full kernel-parity suite passes, including a new randomized ceiling scenario added to `test_kernel_parity.py`'s existing seeded sweep - constructed to derive its values from the sweep's *existing* floor-keep draws rather than call fresh `rng.*()` methods, so the existing seeded scenarios (including car-charging edge cases the suite depends on) are unaffected
- The 20-scenario committed baseline regression (`./run_all --test random`) matches bit-for-bit against the committed baseline, confirming zero behavioural change when the feature is unused

### Also covered

- `execute.py`/`output.py`: plan status markers mirroring the existing floor indicator (`[Manual SoC Max]`, a ⬇ plan icon)
- `plan.py`: the export-window merge guard already applied to the floor (`discard_unused_charge_slots`) is now applied to the ceiling too (`discard_unused_export_slots`), so two export windows can't be silently combined across a ceiling deadline
- `web.py`/`web_helper.py`: the plan page's SOC dropdown gained a second "Maximum SOC" control, reusing the existing `handleSocOverride` JS function (parametrised with an `isMax` flag rather than duplicated)
- `docs/customisation.md`: new section mirroring the existing `manual_soc` writeup

Fixes #1578

## Test plan

- [x] New `test_manual_soc_max.py`: parsing, independence from `manual_soc`, `off` clearing, and the floor/ceiling conflict resolution
- [x] New test in `test_discard_unused_export_slots.py` for the ceiling merge guard
- [x] Extended `test_kernel_parity.py`'s random sweep to exercise the ceiling (RNG-draw-neutral, see above)
- [x] `saving_session`, `manual_soc`, `discard_unused_charge_slots`, `discard_unused_export_slots`, `optimise_levels`, and the 20-scenario baseline regression all pass unchanged
- [x] `./run_pre_commit` passes (black/ruff/cspell/markdownlint)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/predbat/prediction.py [hunk 1]
```diff
@@ -715,6 +718,12 @@ def run_prediction(self, charge_limit, charge_window, export_window, export_limi
                 keep_minute_scaling = max(keep_minute_scaling, 10.0)
                 best_soc_keep = max(best_soc_keep, min(alert_keep / 100.0 * soc_max, soc_max))
 
+            # Soc max keep is a ceiling rather than a floor (e.g. manual_soc_max) - 0 means no ceiling
+            best_soc_max = 0
+            if alert_keep_max > 0:
+                keep_minute_scaling = max(keep_minute_scaling, 10.0)
+                best_soc_max = min(alert_keep_max / 100.0 * soc_max, soc_max)
```
[the remark this answers is not in the document]
Comment (springfall2008) [reply apps/predbat/prediction.py]: _Automated reply from the triage bot._

Confirmed and fixed. `manual_soc_max_value` has `"min": 0, "default": 0` (config.py:1408-1417), and the fetch-side conflict check keeps a 0% ceiling (`floor_value > ceiling` is `0 > 0`, false), so the only thing discarding it was the `> 0` sentinel conflating "no ceiling" with "ceiling of 0%". The default configuration of the feature was therefore a no-op.

Absence is now a negative sentinel instead:

- `prediction.py`: `all_active_keep_max.get(minute_absolute, -1)`, `best_soc_max = -1`, `if alert_keep_max >= 0`, and the metric check is `if best_soc_max >= 0 and soc >= best_soc_max`
- `prediction_kernel.py`: the per-step array now fills with `-1` where no ceiling applies
- `prediction_kernel.cpp`: the mirrored block and the metric check use the same `>= 0` test, and the `alert_keep_max` field comment updated
- `PK_PARITY_REVISION` / `KERNEL_PARITY_REVISION` bumped 11 -> 12, so the prebuilt binaries from the previous CI run are rejected rather than silently disagreeing (verified locally - `kernel_parity` logged `stale binary ... revision 11, expected ... revision 12 - using Python engine`, rebuilt, and passed)

New regression test (T5 in `test_manual_soc_max.py`) runs a real prediction and asserts a 0% ceiling produces a keep penalty, a 50% ceiling produces a smaller one, and no ceiling produces none. Verified it fails against the old code (`ERROR: T5 Expected a keep penalty for holding charge above a 0% ceiling, got 0`) and passes with the fix. `random` (the 20-scenario committed baseline) still matches bit-for-bit, so nothing moves when the feature is unused.

File: apps/predbat/web.py [hunk 2]
```diff
@@ -4773,6 +4779,15 @@ async def html_rate_override(self, request):
                 actual_rate = manual_soc.get(minutes_from_midnight, rate)
                 clear_option = "[{}={}]".format(override_time.strftime("%a %H:%M"), actual_rate)
                 await self.base.async_manual_select("manual_soc", clear_option)
+            elif action == "Set SOC Max":
+                item = self.base.config_index.get("manual_soc_max_value", {})
+                await self.set_state_external(item.get("entity", None), rate)
+                await self.base.async_manual_select("manual_soc_max", selection_option)
+            elif action == "Clear SOC Max":
+                manual_soc_max = self.base.manual_rates("manual_soc_max")
```
[the remark this answers is not in the document]
Comment (springfall2008) [reply apps/predbat/web.py]: _Automated reply from the triage bot._

The `update=False` change itself was already applied in 02f8e6e4, and it matches the `Clear Load` / `Clear SOC` branches above it, so the point stands. However that Autofix commit stripped the line's indentation, leaving `web.py` unparseable (`SyntaxError: Expected dedent, found end of file` at 6290, and black `Cannot parse: 4787:0`) and pre-commit CI red. Indentation restored in this push; `run_pre_commit` passes locally.

File: apps/predbat/tests/test_manual_soc_max.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    orig_log = my_predbat.log
+    my_predbat.log = lambda msg, *args, **kwargs: log_messages.append(str(msg))
+    my_predbat.manual_soc_keep = my_predbat.manual_rates("manual_soc", default_rate=my_predbat.get_arg("manual_soc_value"))
+    my_predbat.manual_soc_max_keep = my_predbat.manual_rates("manual_soc_max", default_rate=my_predbat.get_arg("manual_soc_max_value"))
+
+    # Reproduce the merge fetch_config_options() performs, exercising the actual conflict-resolution code
+    my_predbat.alert_active_keep = {}
+    my_predbat.all_active_keep = my_predbat.alert_active_keep.copy()
+    for minute, soc_value in my_predbat.manual_soc_keep.items():
+        my_predbat.all_active_keep[minute] = max(my_predbat.all_active_keep.get(minute, 0), soc_value)
```
[the remark this answers is not in the document]
Comment (springfall2008) [reply apps/predbat/tests/test_manual_soc_max.py]: _Automated reply from the triage bot._

Correct - T4 copy-pasted the merge loop, so the comment claiming it "exercises the actual conflict-resolution code" was false and the test would have passed with the fetch.py block deleted.

Taken the extract-a-helper option: the merge and conflict resolution are now `Fetch.combine_active_keep()` (fetch.py), called from `fetch_sensor_data` where the block used to sit, and T4 calls that same method instead of a copy. The duplicated loop is gone from the test.

Impact analysis on `fetch_sensor_data` reports HIGH risk (it is step 1 of `update_pred`, `run_time_loop`, `update_time_loop` and `initialize`), so for the record: the extraction is a straight lift of a self-contained block, no signature change and no reordering relative to its neighbours, and the quick suite plus `manual_soc`, `manual_soc_max`, `discard_unused_export_slots`, `model`, `kernel_parity` and the `random` baseline all pass.
